### PR TITLE
Separate View Evaluation and Overrides

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
This simplifies the View Evaluation by separating it from the logic of
overriding CreatePCollectionView.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @jkff

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  private transient ViewOverrideFactory<Integer, List<Integer>> factory =
+      new ViewOverrideFactory<>();
+
+  @Test
+  public void replacementSucceeds() {
+    PCollection<Integer> ints = p.apply("CreateContents", Create.of(1, 2, 3));
+    final PCollectionView<List<Integer>> view =
+        PCollectionViews.listView(p, WindowingStrategy.globalDefault(), ints.getCoder());
+    PCollectionView<List<Integer>> afterReplacement =
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java]: This variable is unused; are you sure this test is testing the code you changed?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    PTransform<PCollection<Integer>, PCollectionView<List<Integer>>> replacement =
+        factory.getReplacementTransform(CreatePCollectionView.<Integer, List<Integer>>of(view));
+    ints.apply(replacement);
+    final AtomicBoolean writeViewVisited = new AtomicBoolean();
+    p.traverseTopologically(
+        new PipelineVisitor.Defaults() {
+          @Override
+          public void visitPrimitiveTransform(Node node) {
+            assertThat(
+                "WriteView should be the last primtitive in the graph",
```
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java]: * typo: primitive
* why?
* this condition is actually testing that there's exactly one WriteView transform, not that it's "last"

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java]: Yes, but I've made this more explicit. The application is the important part; we're creating a view manually to get around the `View.asFoo` transforms.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java (hunk 2 above)
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/ViewOverrideFactoryTest.java]: * Fixed
* Updated to assert that we only write the view once, and it has the output we expect.

**Pull request closed without merging**